src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -346,34 +345,34 @@ Plan buildPlan()
         rowsIteration = planFactory.recheckFilter(command.rowFilter(), rowsIteration);
         rowsIteration = planFactory.limit(rowsIteration, command.limits().rows());
 
-        Plan optimizedPlan;
-        optimizedPlan = QUERY_OPT_LEVEL > 0
-                        ? rowsIteration.optimize()
-                        : rowsIteration;
-        optimizedPlan = KeyRangeIntersectionIterator.INTERSECTION_CLAUSE_LIMIT > 0 && QUERY_OPT_LEVEL <= 1
-                        ? optimizedPlan.limitIntersectedClauses(KeyRangeIntersectionIterator.INTERSECTION_CLAUSE_LIMIT)
-                        : optimizedPlan;
+        // We apply the intersection clause limit first before optimizing so we reduce the size of the
+        // plan given to the optimizer and hence we reduce the plan search space and speed up optimization.
+        // It is possible that some index operators like ':' expand to a huge number of MATCH predicates
+        // (see CNDB-10085) and could overload the optimizer.
+        Plan plan = rowsIteration.limitIntersectedClauses(KeyRangeIntersectionIterator.INTERSECTION_CLAUSE_LIMIT);
+        if (QUERY_OPT_LEVEL > 0)
+            plan = plan.optimize();
```

Comment:
What if we limited it to a multiple of `INTERSECTION_CLAUSE_LIMIT`? Or perhaps we could sort the `rowsIteration` based on expected selectivity by inspecting the search term and the associated operator? I am concerned that this will negatively impact other queries.

Also, @jbellis is looking into pushing the analyzer logic into a single expression that is then run within an index. I think that should also reduce the complexity observed in this plan and let us only truncate the results for analyzed queries.

Reply:
Thanks, I missed that it was already sorted. It looks good.